IHP/RouterSupport.hs

```diff
@@ -68,9 +69,81 @@ import qualified IHP.WebSocket as WS
+    pure ioRouteResult'
+
+applyContextSetter :: (TMap.TMap -> TMap.TMap) -> ControllerContext -> IO ControllerContext
+applyContextSetter setter ctx@ControllerContext { customFieldsRef } = do
+    modifyIORef customFieldsRef (applySetter setter)
+    pure $ ctx { customFieldsRef }
+    where
+        fromSetter :: (TMap.TMap -> TMap.TMap) -> TMap.TMap
+        fromSetter f = f TMap.empty
+
+        applySetter :: (TMap.TMap -> TMap.TMap) -> TMap.TMap -> TMap.TMap
+        applySetter f map = TMap.union (fromSetter f) map
+
+runAction'
+    :: forall application controller
+     . ( Controller controller
+       , ?applicationContext :: ApplicationContext
+       , ?context :: RequestContext
+       , InitControllerContext application
+       , ?application :: application
+       , Typeable application
+       , Typeable controller
+       )
+     => controller -> (TMap.TMap -> TMap.TMap) -> IO ResponseReceived
+runAction' controller contextSetter = do
+    let ?modelContext = ApplicationContext.modelContext ?applicationContext
+    let ?requestContext = ?context
+    contextOrErrorResponse <- newContextForAction controller
+    case contextOrErrorResponse of
+        Left res -> res
+        Right context -> do
+            context' <- applyContextSetter contextSetter context
+            runActionWithContext context' controller
+
+type RouteParseResult = IO (TMap.TMap -> TMap.TMap, (TMap.TMap -> TMap.TMap) -> IO ResponseReceived)
+type RouteParser = Parser (RouteParseResult)
+
+toRouteParser :: Parser (IO ResponseReceived) -> RouteParser
+toRouteParser parser = do
+    controller <- parser
+    pure $ pure (\t -> t, \_ -> controller)
+
+toRouteParser' :: Parser ((TMap.TMap -> TMap.TMap) -> IO ResponseReceived) -> RouteParser
+toRouteParser' parser = do
+    controller <- parser
+    pure $ pure (\t -> t, controller)
+
+toRouteParseResult :: IO ResponseReceived -> RouteParseResult
+toRouteParseResult ioResponseReceived = pure (\t -> t, \_ -> ioResponseReceived)
 
 class FrontController application where
-    controllers :: (?applicationContext :: ApplicationContext, ?application :: application, ?context :: RequestContext) => [Parser (IO ResponseReceived)]
+    controllers
+        :: forall controller
+         . (?applicationContext :: ApplicationContext, ?application :: application, ?context :: RequestContext)
+        => [RouteParser]
+
+    router
+        :: (?applicationContext :: ApplicationContext, ?application :: application, ?context :: RequestContext)
+        => [RouteParser] -> RouteParser
+    router = defaultRouter
+
+    defaultRouter
+        :: (?applicationContext :: ApplicationContext, ?application :: application, ?context :: RequestContext)
+        => [RouteParser] -> RouteParser
+    defaultRouter additionalControllers = do
+        let allControllers = controllers <> additionalControllers
+        ioResponseReceived <- choice $ map (\r -> r <* endOfInput) allControllers
+        pure ioResponseReceived
```

Comment:
It seems like this could be moved out of the type class into a standalone function